src/MutationFunctionHandler.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * Returns a function that, when called with arguments,
+ * extends the path with mutationExpressions.
+ *
+ * It uses the current path expression as domain
+ * and the given arguments as range.
+ * These arguments can either be raw, or other path expressions.
+ *
+ * Requires:
+ * - a pathExpression property on the path proxy and all non-raw arguments.
+ */
+export default class MutationFunctionHandler {
+  constructor(mutationType, allowZeroArgs) {
+    this._mutationType = mutationType;
+    this._allowZeroArgs = allowZeroArgs;
+  }
+
+  execute(path, proxy) {
+    return (...args) => {
+      const mutationExpressions = {
+        then: (resolve, reject) =>
+          this.createMutationExpressions(path, proxy, args).then(resolve, reject),
+      };
+      return path.extend({ mutationExpressions });
+    };
+  }
+
+  async createMutationExpressions(path, proxy, args) {
+    // Check if the given arguments are valid
+    if (!this._allowZeroArgs && !args.length)
+      throw new Error(`Mutation on ${path} can not be invoked without arguments`);
```

Comment:
I would leave it in, because we interpret "no args" as "all possible values".
For example, `.delete()` deletes all possible values.
This is however only possible for deletions, additions will (and should) throw this error, because `.add()` does not really make sense.

Will make it sync in any case.